File: 5lista/zad3/PrioQueue.java
import java.util.Arrays;

public class PrioQueue {
    final static int EMPTY_QUEUE = -2147483647;
    Element[] heap;
    int size;

    class Element {
        int val;
        float prio;
    
        public Element(int val, float prio) {
            this.val = val;
            this.prio = prio;
        }
    }

    public PrioQueue(int size) {
        this.heap = new Element[size];
        this.size = 0;
    }

    public int empty() {
        if(heap[0] == null) {
            return 0;
        } else {
            return 1;
        }
    }

    private void resizeHeap() {
        Element[] newHeap = Arrays.copyOf(this.heap, this.size*2);
        this.heap = newHeap;
    }

    public void insert(int x, float p) {
        if(size >= this.heap.length) {
            resizeHeap();
        }
        int i = this.size;
        int j = (i-1) / 2;
        Element newElement = new Element(x, p);
        this.size++;

        while(i > 0 && this.heap[j].prio > p) {
            this.heap[i] = this.heap[j];
            i = j;
            j = (i-1) / 2;
        }
        this.heap[i] = newElement;
    }

    public int pop() {
        if(this.size == 0) {
            return EMPTY_QUEUE;
        } else {
            int val = this.heap[0].val;
            this.size--;
            if(this.size == 0) {
                this.heap[0] = null;
                return val;
            }
            Element last = this.heap[this.size];
            this.heap[this.size] = null;

            int i = 0;
            int j = 1;

            while(j < this.size) {
                if(j+1 < this.size && this.heap[j+1].prio < this.heap[j].prio) {
                    j++;
                }
                if(last.prio <= this.heap[j].prio) {
                    break;
                }
                this.heap[i] = this.heap[j];
                i = j;
                j = 2 * j + 1;
            }
            this.heap[i] = last;
            return val;
        }
    }

    public String top() {
        if(this.size > 0) {
            return "" + this.heap[0].val;
        } else {
            return "";
        }
    }

    public void priority(int x, float p) {
        findElements(x, p, 0);
    }

    private void findElements(int x, float p, int idx) {
        if(this.size > idx && this.heap[idx] != null) {
            if(this.heap[idx].val == x && this.heap[idx].prio > p) {
                this.heap[idx].prio = p;
                fixElement(idx);
            }
            int left = 2 * idx + 1;
            findElements(x, p, left);
            int right = 2 * idx + 2;
            findElements(x, p, right);
        }
    }

    private void fixElement(int idx) {
        Element toFix = this.heap[idx];

        int i = idx;
        int j = (idx - 1) / 2;

        while(j >= 0 && this.heap[j].prio > toFix.prio ) {
            this.heap[i] = this.heap[j];
            i = j;
            j = (i - 1) / 2;
            if(j == 0) {
                break;
            }
        }
        this.heap[i] = toFix;
    }

    public String print() {
        String s = "";
        Element[] arr = Arrays.copyOf(this.heap, this.size);
        int size= this.size-1;

        while(size >= 0) {
            s += "(" + arr[0].val + ", " + arr[0].prio + ") ";
            Element last = arr[size];
            arr[size] = null;

            int i = 0;
            int j = 1;

            while(j < size) {
                if(j+1 < size && arr[j+1].prio < arr[j].prio) {
                    j++;
                }
                if(last.prio <= arr[j].prio) {
                    break;
                }
                arr[i] = arr[j];
                i = j;
                j = 2 * j + 1;
            }
            arr[i] = last;
            size--;
        }
        return s;
    }
}

File: 5lista/python/prioQueue.py
import fileinput

class Element:
    def __init__(self, val, prio):
        self.val = val
        self.prio = prio

class PrioQueue:
    def __init__(self):
        self.size = 0
        self.heap = []

    def empty(self):
        if(len(self.heap) == 0):
            return True
        else:
            return False
    
    def insert(self, x, p):
        i = self.size
        j = (i-1)//2
        newElement = Element(x, p)
        self.heap.append(newElement)
        self.size += 1
        self.fixElements(self.size-1)
        # while i > 0 and self.heap[j].prio > p:
        #     self.heap[i] = self.heap[j]
        #     i = j
        #     j = (i-1)//2
        # self.heap[i] = newElement
    
    def pop(self):
        # print("size: ", self.size)
        if self.size == 0:
            return None
            
        val = self.heap[0].val
        self.size -= 1

        lastElement = self.heap[self.size]
        del self.heap[self.size]
        if self.size == 0:
            return val
        i = 0
        j = 1
        while j < self.size:
            if j+1 < self.size and self.heap[j+1].prio < self.heap[j].prio:
                j += 1
            if lastElement.prio <= self.heap[j].prio:
                break
            self.heap[i] = self.heap[j]
            i = j
            j = 2*j+1
        self.heap[i] = lastElement
        return val

    def top(self):
        if self.size > 0:
            return self.heap[0].val
        else:
            return None
    
    def priority(self, x, p):
        self.findElements(x, p, 0)

    def findElements(self, x, p, idx):
        if self.size > idx:
            if self.heap[idx].val == x and self.heap[idx].prio > p:
                self.heap[idx].prio = p
                self.fixElements(idx)
            left = 2 * idx + 1
            self.findElements(x, p, left)
            right = 2 * idx + 2
            self.findElements(x, p, right)
    
    def fixElements(self, idx):
        toFix = self.heap[idx]
        i = idx
        j = (idx-1)//2

        while j >= 0 and self.heap[j].prio > toFix.prio:
            self.heap[i] = self.heap[j]
            i = j
            j = (i-1)//2
            if j == 0:
                break
        self.heap[i] = toFix

    def print(self):
        str = ""
        arr = list(self.heap)
        size = self.size-1

        while size >= 0:
            # val = str(arr[0].val)
            # prio = str(arr[0].prio)
            str += "(" + '{}'.format(arr[0].val) + ", " + '{}'.format(arr[0].prio) + ") "
            lastElement = arr[size]
            arr[size] = None

            i = 0
            j = 1

            while j < size:
                if j+1 < size and arr[j+1].prio < arr[j].prio:
                    j += 1
                if lastElement.prio <= arr[j].prio:
                    break
                arr[i] = arr[j]
                i = j
                j = 2*j+1
            arr[i] = lastElement
            size -= 1
        print(str)
        return str

    def make(self, lines):
        if lines[0] == "insert":
            self.insert(int(lines[1]), float(lines[2]))
        elif lines[0] == "empty":
            if self.empty():
                print(0)
            else:
                print(1)
        elif lines[0] == "top":
            if self.empty():
                print()
            else:
                print(self.top())
        elif lines[0] == "pop":
            if self.empty():
                print()
            else:
                print(self.pop())
        elif lines[0] == "priority":
            self.priority(int(lines[1]), float(lines[2]))
        elif lines[0] == "print":
            self.print()
        else:
            print("Wrong arg!")


pq = PrioQueue()
for line in fileinput.input():
    lines = line.split()
    pq.make(lines)
    
# pq = PrioQueue()
# pq.insert(10, 0)
# pq.insert(6, 3.12)
# pq.insert(13, 8.5)
# pq.insert(24, 2.7)
# pq.insert(100, 0)
# pq.insert(13, 16)
# pq.insert(10, 6)
# pq.insert(10, 5)
# pq.print()

# print(pq.pop())
# print(pq.pop())
# pq.print()

# pq.priority(10, 0)
# pq.print()

# print(pq.pop())
# print(pq.pop())
# print(pq.pop())
# print(pq.pop())
# print(pq.pop())
# print(pq.pop())
# print(pq.pop())
# print(pq.pop())